src/ipfinder.cpp

```diff
@@ -0,0 +1,163 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "ipfinder.h"
+#include "constants.h"
+#include "errorhandler.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+#include "settingsholder.h"
+
+#include <QHostAddress>
+#include <QHostInfo>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+#include <QUrl>
+
+namespace {
+Logger logger(LOG_NETWORKING, "IPFinder");
+}
+
+IPFinder::IPFinder(QObject* parent) : QObject(parent) {
+  MVPN_COUNT_CTOR(IPFinder);
+}
+
+IPFinder::~IPFinder() {
+  MVPN_COUNT_DTOR(IPFinder);
+  if (m_lookupId > -1) {
+    QHostInfo::abortHostLookup(m_lookupId);
+  }
+}
+
+void IPFinder::start() {
+  logger.log() << "Starting the ip-lookup";
+
+  QUrl url(Constants::API_URL);
+  m_lookupId = QHostInfo::lookupHost(url.host(), this,
+                                     SLOT(dnsLookupCompleted(QHostInfo)));
+}
+
+void IPFinder::dnsLookupCompleted(const QHostInfo& hostInfo) {
+  logger.log() << "DNS lookup completed";
+
+  m_lookupId = -1;
+
+  if (hostInfo.error() != QHostInfo::NoError) {
+    logger.log() << "Unable to perform a DNS Lookup:" << hostInfo.errorString();
+    emit completed(QString(), QString(), QString());
+    deleteLater();
+    return;
+  }
+
+  bool ipv6Enabled = SettingsHolder::instance()->ipv6Enabled();
+
+  for (const QHostAddress& address : hostInfo.addresses()) {
```

Comment:
This isn't necessarily an issue because we control the DNS records for the API, but it's not uncommon to get a long list of IP addresses back when resolving a hostname, but we only really need a single IPv4 and IPv6 address to complete the IP check.